all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,454 @@
+* Ability to provide environment variables to represent values in the configuration
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+* Profile names within the configuration
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
+* File-based hierarchy. We don't need many levels of file merging for these options at this time.
+* Extreme focus on human authoring of this format. This is because it'd likely go against the "0 new dependencies" rule.
+* Use external configuration by default. Users need to opt-in to this for compatibility and clarity reasons (but it is
+  of course easy).
+
+### High-level Usage Examples
+
+#### Simple Switch from Local to Cloud
+
+Similar to `temporal env` today, you can use the CLI to manage configurations with `temporal config`.
+
+For example, say you had this Go code for starting a workflow:
+
+```go
+options, err := client.LoadOptionsFromConfig("default")
+if err != nil {
+  return err
+}
+cl, err := client.Dial(options)
+run, err := cl.ExecuteWorkflow(
+  ctx,
+  client.StartWorkflowOptions{ID: "my-id", TaskQueue: "my-task-queue"},
+  MyWorkflow,
+)
+```
+
+By default this only works on `localhost:7233` with namespace `default` when you run it. But if you were a cloud user,
+you could do this in environment variables for mTLS auth:
+
+    export TEMPORAL_ADDRESS=my-ns.a1b2c.tmprl.cloud:7233
+    export TEMPORAL_NAMESPACE=my-ns.a1b2c
+    export TEMPORAL_TLS_CLIENT_CERT_PATH=path/to/my/client.pem
+    export TEMPORAL_TLS_CLIENT_KEY_PATH=path/to/my/client.key
+
+Now running the same code again unmodified will run against cloud. You can do ths same thing with config instead of
+environment variables:
+
+    temporal config set --key address --value my-ns.a1b2c.tmprl.cloud:7233
+    temporal config set --key namespace --value my-ns.a1b2c
+    temporal config set --key tls.clientCertPath --value path/to/my/cert.pem
+    temporal config set --key tls.clientKeyPath --value path/to/my/cert.key
+
+Now running the same code again unmodified will run against cloud. Or in a Temporal future with API keys and a
+known/fixed API endpoint:
+
+    temporal config set --key cloudApiKey --value my-api-key
+
+And that would be all that is needed (but that is a future scenario, note `cloudApiKey` is not a defined key today, but
+`apiKey` is, but you have to put in the `address` and `namespace` still).
+
+#### Profile Switching
+
+Similar to AWS tooling and other tools, you can have profiles. So you can have:
+
+    temporal config set --profile dev --key address --value my-dev-ns.a1b2c.tmprl.cloud:7233
+    temporal config set --profile dev --key namespace --value my-dev-ns.a1b2c
+    temporal config set --profile dev --key tls.clientCertPath --value path/to/my/dev-cert.pem
+    temporal config set --profile dev --key tls.clientKeyPath --value path/to/my/dev-cert.key
```

Comment:
It would be nicer if users weren't required to run imperative commands to do config for customers who lack such hooks.  I.e. they could just declare the values statically and then the SDK can parse that.

I also see a lot of "it worked on my box" bugs with these sorts of imperative configs where somebody has already run script A and then runs script B, which works fine.  Whereas another dev checks out the code and only runs script B and it doesn't work.

Reply:
Yes, the advantage of  configuration files is that they simply need to be present--nothing needs to be explicitly run beforehand.  Thus we eliminate a class of bugs where the script didn't get run.